File: golang/cmd/day01/Day01.go
package main

import (
	"bytes"
	"fmt"
	"golang.org/x/exp/maps"
	"nutgaard/aoc/internal/utils"
	"strings"
)

var digitMap = make(map[string]rune)

func main() {
	testInput1 := utils.ReadFile("cmd/day01/Part01_test.txt")
	testInput2 := utils.ReadFile("cmd/day01/Part02_test.txt")
	input := utils.ReadFile("cmd/day01/Input.txt")

	digitMap["one"] = '1'
	digitMap["two"] = '2'
	digitMap["three"] = '3'
	digitMap["four"] = '4'
	digitMap["five"] = '5'
	digitMap["six"] = '6'
	digitMap["seven"] = '7'
	digitMap["eight"] = '8'
	digitMap["nine"] = '9'

	test := decodeDigits("zoneight234")
	fmt.Println(test)

	utils.Verify(testInput1, 142, part1)
	utils.Verify(testInput2, 281, part2)

	utils.Run("part1", part1, input)
	utils.Run("part2", part2, input)
}

func part1(input string) int {
	lines := strings.Split(input, "\n")
	sum := 0
	for _, line := range lines {
		sum += calibrationValue(line)
	}
	return sum
}
func part2(input string) int {
	lines := strings.Split(input, "\n")
	sum := 0
	for _, line := range lines {
		sum += calibrationValue(decodeDigits(line))
	}
	return sum
}

func isDigit(u rune) bool {
	return u >= '0' && u <= '9'
}

func calibrationValue(line string) int {
	n := len(line)
	v := 0
	for i := 0; i < n; i++ {
		ch := rune(line[i])
		if isDigit(ch) {
			v = 10 * int(ch-'0')
			break
		}
	}

	for i := 0; i < n; i++ {
		ch := rune(line[n-1-i])
		if isDigit(ch) {
			v += int(ch - '0')
			break
		}
	}

	return v
}

func decodeDigits(input string) string {
	var b bytes.Buffer
	keys := maps.Keys(digitMap)

	for i := 0; i < len(input); i++ {
		ch := rune(input[i])
		if isDigit(ch) {
			b.WriteRune(ch)
		} else {
			for _, key := range keys {
				if strings.HasPrefix(input[i:], key) {
					b.WriteRune(digitMap[key])
				}
			}
		}
	}

	return b.String()
}

File: src/day01/Day01.kt
package day01

import println
import readInput
import timed
import verifySolution

const val dir = "day01"

fun main() {
    val testInputPart1 = readInput("$dir/Part01_test")
    val testInputPart2 = readInput("$dir/Part02_test")

    verifySolution("zoneight", "18", ::decodeDigits)
    verifySolution("zoneight234", "18234", ::decodeDigits)
    verifySolution(testInputPart1, 142, ::part1)
    verifySolution(testInputPart2, 281, ::part2)

    val input = readInput("$dir/Input")
    timed("part1") { part1(input).println() }
    timed("part2") { part2(input).println() }
}

fun part1(input: List<String>): Int {
    return input.map(::calibrationValue).sum()
}

fun part2(input: List<String>): Int {
    return input.map(::decodeDigits).map(::calibrationValue).sum()
}

fun calibrationValue(line: String): Int {
    val first = line.first { it.isDigit() }
    val last = line.last { it.isDigit() }

    return "$first$last".toInt()
}

val digitMap = mapOf(
    "one" to 1,
    "two" to 2,
    "three" to 3,
    "four" to 4,
    "five" to 5,
    "six" to 6,
    "seven" to 7,
    "eight" to 8,
    "nine" to 9,
)
val digitMapRegexp = Regex("^(?:${digitMap.keys.joinToString("|")})")

fun decodeDigits(line: String): String {
    return buildString {
        line.forEachIndexed { index, c ->
            if (c.isDigit()) {
                append(c)
            } else {
                val substring = line.substring(index)
                val digitMatch = digitMapRegexp.find(substring, 0)?.value
                if (digitMatch != null) {
                    append(digitMap[digitMatch])
                }
            }
        }
    }
}